sg/initfile.go

```diff
@@ -33,74 +37,78 @@ func generateInitFile(g *codegen.File, pkg *doc.Package, mks []Makefile) error {
-						g.P("arg", i, " := args[", i, "]")
-					case intType:
-						g.P("arg", i, ", err :=", g.Import("strconv"), ".Atoi(args[", i, "])")
-						g.P("if err != nil {")
-						g.P(`logger.Fatalf("can't convert argument %q to int", args[`, i, `])`)
-						g.P("}")
-					case boolType:
-						g.P("arg", i, ", err :=", g.Import("strconv"), ".ParseBool(args[", i, "])")
-						g.P("if err != nil {")
-						g.P(`logger.Fatalf("can't convert argument to bool", args[`, i, `])`)
-						g.P("}")
-					}
-					i++
-				}
-			}
-			ns := "{}."
-			for _, mk := range mks {
-				if mk.namespaceName() != "" && mk.namespaceName() == function.Recv {
-					ns = getNamespaceMetadata(mk)
+		for _, mk := range mks {
+			if mk.namespaceName() == function.Recv {
+				g.P(`case "`, getTargetFunctionName(function), `":`)
+				loggerName := getTargetFunctionName(function)
+				// Remove namespace from loggerName
+				if strings.Contains(loggerName, ":") {
+					loggerName = strings.Split(loggerName, ":")[1]
 				}
-			}
-			g.P(
-				"err = ",
-				strings.ReplaceAll(getTargetFunctionName(function), ":", ns),
-				"(ctx,",
-				strings.Join(args, ","),
-				")",
-			)
-			g.P("if err != nil {")
-			g.P("logger.Fatal(err)")
-			g.P("}")
-		} else {
-			ns := "{}."
-			for _, mk := range mks {
-				if mk.namespaceName() != "" && mk.namespaceName() == function.Recv {
-					ns = getNamespaceMetadata(mk)
+				g.P("logger := ", g.Import("go.einride.tech/sage/sg"), ".NewLogger(\"", loggerName, "\")")
+				g.P("ctx = ", g.Import("go.einride.tech/sage/sg"), ".WithLogger(ctx, logger)")
+				if len(function.Decl.Type.Params.List) > 1 {
+					expected := countParams(function.Decl.Type.Params.List) - 1
+					g.P("if len(args) != ", expected, " {")
+					g.P(
+						`logger.Fatalf("wrong number of arguments to %s, got %v expected %v",`,
+						strconv.Quote(getTargetFunctionName(function)), ",",
+						expected, ",",
+						`len(args))`,
+					)
+					g.P(g.Import("os"), ".Exit(1)")
+					g.P("}")
+					var args []string
+					var i int
+					for _, customParam := range function.Decl.Type.Params.List[1:] {
+						for range customParam.Names {
+							args = append(args, fmt.Sprintf("arg%v", i))
+							switch fmt.Sprint(customParam.Type) {
+							case stringType:
+								g.P("arg", i, " := args[", i, "]")
+							case intType:
+								g.P("arg", i, ", err :=", g.Import("strconv"), ".Atoi(args[", i, "])")
+								g.P("if err != nil {")
+								g.P(`logger.Fatalf("can't convert argument %q to int", args[`, i, `])`)
+								g.P("}")
+							case boolType:
+								g.P("arg", i, ", err :=", g.Import("strconv"), ".ParseBool(args[", i, "])")
+								g.P("if err != nil {")
+								g.P(`logger.Fatalf("can't convert argument to bool", args[`, i, `])`)
+								g.P("}")
+							}
+							i++
+						}
+					}
+					ns := "{}."
+					for _, mk := range mks {
+						if mk.namespaceName() != "" && mk.namespaceName() == function.Recv {
+							ns = getNamespaceMetadata(mk)
+						}
+					}
```

Comment:
Should this be a loop at all? We are already inside an mk so now we are looping twice. Maybe

```suggestion
		if mk.namespaceName() != "" {
			ns = getNamespaceMetadata(mk)
		}
```

Reply:
I refactored this before I saw your comment :)